File: cpp/src/strong_dis.cpp
#include "strong_dis.hpp"

#include <cmath>
#include <numeric>

namespace {

double column_sum(const std::vector<double>& matrix, std::size_t n, std::size_t column) {
    double sum = 0.0;
    for (std::size_t row = 0; row < n; ++row) {
        sum += matrix[row * n + column];
    }
    return sum;
}

std::vector<double> multiply_square(
    const std::vector<double>& left,
    const std::vector<double>& right,
    std::size_t n
) {
    std::vector<double> product(n * n, 0.0);

    for (std::size_t i = 0; i < n; ++i) {
        for (std::size_t k = 0; k < n; ++k) {
            const double left_value = left[i * n + k];
            if (left_value == 0.0) {
                continue;
            }
            for (std::size_t j = 0; j < n; ++j) {
                product[i * n + j] += left_value * right[k * n + j];
            }
        }
    }

    return product;
}

std::vector<double> multiply_matrix_vector(
    const std::vector<double>& matrix,
    const std::vector<double>& vector,
    std::size_t n
) {
    std::vector<double> result(n, 0.0);

    for (std::size_t i = 0; i < n; ++i) {
        double sum = 0.0;
        for (std::size_t j = 0; j < n; ++j) {
            sum += matrix[i * n + j] * vector[j];
        }
        result[i] = sum;
    }

    return result;
}

} // namespace

bool strong_dis_mm(const std::vector<double>& f, std::size_t n, std::size_t a, std::size_t b) {
    auto m = f;

    for (std::size_t iter = 1; iter <= 2 * n - 1; ++iter) {
        const double ma = column_sum(m, n, a);
        const double mb = column_sum(m, n, b);

        if ((iter % 2 != 0 && ma < mb) || (iter % 2 == 0 && mb < ma)) {
            return true;
        }
        if (ma != mb) {
            return false;
        }

        m = multiply_square(m, f, n);
    }

    return false;
}

bool strong_dis_mv(const std::vector<double>& f, std::size_t n, std::size_t a, std::size_t b) {
    std::vector<double> v1(n, 0.0);
    std::vector<double> v2(n, 0.0);
    v1[a] = 1.0;
    v2[b] = 1.0;

    for (std::size_t iter = 1; iter <= 2 * n; ++iter) {
        v1 = multiply_matrix_vector(f, v1, n);
        v2 = multiply_matrix_vector(f, v2, n);

        const double sum_v1 = std::accumulate(v1.begin(), v1.end(), 0.0);
        const double sum_v2 = std::accumulate(v2.begin(), v2.end(), 0.0);

        if (std::abs(sum_v1 - sum_v2) > 1e-9) {
            if (iter % 2 != 0) {
                return sum_v1 <= sum_v2;
            }
            return sum_v1 > sum_v2;
        }
    }

    return false;
}

File: rust/src/strong_dis.rs
pub fn strong_dis_mm(f: &[f64], n: usize, a: usize, b: usize) -> bool {
    let mut m = f.to_vec();

    for iter in 1..=(2 * n - 1) {
        let ma = column_sum(&m, n, a);
        let mb = column_sum(&m, n, b);

        if (iter % 2 != 0 && ma < mb) || (iter % 2 == 0 && mb < ma) {
            return true;
        }
        if ma != mb {
            return false;
        }

        m = multiply_square(&m, f, n);
    }

    false
}

pub fn strong_dis_mv(f: &[f64], n: usize, a: usize, b: usize) -> bool {
    let mut v1 = vec![0.0; n];
    let mut v2 = vec![0.0; n];
    v1[a] = 1.0;
    v2[b] = 1.0;

    for iter in 1..=(2 * n) {
        v1 = multiply_matrix_vector(f, &v1, n);
        v2 = multiply_matrix_vector(f, &v2, n);

        let sum_v1: f64 = v1.iter().sum();
        let sum_v2: f64 = v2.iter().sum();

        if (sum_v1 - sum_v2).abs() > 1e-9 {
            if iter % 2 != 0 {
                return sum_v1 <= sum_v2;
            }
            return sum_v1 > sum_v2;
        }
    }

    false
}

fn column_sum(matrix: &[f64], n: usize, column: usize) -> f64 {
    (0..n).map(|row| matrix[row * n + column]).sum()
}

fn multiply_square(left: &[f64], right: &[f64], n: usize) -> Vec<f64> {
    let mut product = vec![0.0; n * n];

    for i in 0..n {
        for k in 0..n {
            let left_value = left[i * n + k];
            if left_value == 0.0 {
                continue;
            }
            for j in 0..n {
                product[i * n + j] += left_value * right[k * n + j];
            }
        }
    }

    product
}

fn multiply_matrix_vector(matrix: &[f64], vector: &[f64], n: usize) -> Vec<f64> {
    let mut result = vec![0.0; n];

    for i in 0..n {
        let mut sum = 0.0;
        for j in 0..n {
            sum += matrix[i * n + j] * vector[j];
        }
        result[i] = sum;
    }

    result
}

#[cfg(test)]
mod tests {
    use super::{strong_dis_mm, strong_dis_mv};

    #[test]
    fn mm_and_mv_return_true_for_simple_asymmetric_case() {
        let f = vec![0.0, 1.0, 0.0, 0.0];

        assert!(strong_dis_mm(&f, 2, 0, 1));
        assert!(strong_dis_mv(&f, 2, 0, 1));
    }

    #[test]
    fn mm_and_mv_return_false_when_arguments_are_equal() {
        let f = vec![1.0, 0.0, 0.0, 1.0];

        assert!(!strong_dis_mm(&f, 2, 0, 0));
        assert!(!strong_dis_mv(&f, 2, 0, 0));
    }
}
